src/frontend/mod.rs

```diff
@@ -0,0 +1,153 @@
+#![allow(non_snake_case)]
+
+use crate::common::Scores;
+use chat::Chat;
+use dioxus::prelude::*;
+use std::sync::{Arc, Mutex};
+use wasm_bindgen::prelude::*;
+use web_sys::console;
+use web_sys::WebSocket;
+
+mod chat;
+
+#[wasm_bindgen(start)]
+pub fn run_app() {
+    launch(App);
+}
+
+#[derive(Clone, Default)]
+struct State {
+    inner: Arc<Mutex<InnerState>>,
+}
+
+#[derive(Default)]
+struct InnerState {
+    scores: Option<Scores>,
+    socket: Option<WebSocket>,
+}
+
+impl State {
+    fn set_scores(&self, scores: Scores) {
+        self.inner.lock().unwrap().scores = Some(scores);
+    }
+
+    fn scores(&self) -> Option<Scores> {
+        self.inner.lock().unwrap().scores
+    }
+
+    fn set_socket(&self, socket: WebSocket) {
+        self.inner.lock().unwrap().socket = Some(socket);
+    }
+
+    fn send_message(&self, msg: &str) -> bool {
+        if let Some(socket) = &self.inner.lock().unwrap().socket {
+            let _ = socket.send_with_str(msg);
+            true
+        } else {
+            log_to_console("attempted to send msg without a socket configured");
+            false
+        }
+    }
+}
+
+fn scores_from_formdata(form: &FormData) -> Option<Scores> {
+    let data = form.values();
+
+    let o: f32 = data.get("o")?.as_value().parse().ok()?;
+    let c: f32 = data.get("c")?.as_value().parse().ok()?;
+    let e: f32 = data.get("e")?.as_value().parse().ok()?;
+    let a: f32 = data.get("a")?.as_value().parse().ok()?;
+    let n: f32 = data.get("n")?.as_value().parse().ok()?;
+
+    if !(0. ..=100.).contains(&o) {
+        return None;
+    }
+    if !(0. ..=100.).contains(&c) {
+        return None;
+    }
+    if !(0. ..=100.).contains(&e) {
+        return None;
+    }
+    if !(0. ..=100.).contains(&a) {
+        return None;
+    }
+    if !(0. ..=100.).contains(&n) {
+        return None;
+    }
+
+    Some(Scores { o, c, e, a, n })
+}
```

Comment:
Can't this be done by implementing `impl TryFrom<FormData> for Score`?

Reply:
sure, but maybe in another PR ? this one is kinda big already so let's not introduce more complexities than necessary

Reply:
Alright, I'll do it in another pr